pkg/alert/alert.go

```diff
@@ -0,0 +1,534 @@
+	deleteOutdatedAlertDocsPayload := fmt.Sprintf(`{ "query": { "terms" : { "testContext.keyword" : ["%s"] } } }`, strings.Join(testNames,`", "`))
+	if err := alert.elasticRequest("/tm-alert*/_delete_by_query", http.MethodPost, deleteOutdatedAlertDocsPayload, nil); err != nil {
+		return errors.Wrapf(err, "failed to delete recovered elasticsearch test docs")
+	}
+	return nil
+}
+
+//splitSlackMessage split message line wise based on given characters limit
+func splitSlackMessage(message string, charactersLimit int) []string {
+	var messageSplits []string
+	lines := strings.Split(message, "\n")
+	messageSplitItem := ""
+	for _, line := range lines {
+		if len(messageSplitItem)+len(line) < charactersLimit {
+			messageSplitItem += "\n" + line
+		} else {
+			messageSplits = append(messageSplits, messageSplitItem)
+			messageSplitItem = line
+		}
+	}
+	messageSplits = append(messageSplits, messageSplitItem)
+	return messageSplits
+}
+
+//elasticRequest send HTTP request to elasticsearch
+func (alert *Alert) elasticRequest(urlAttributes, httpMethod, payloadFormated string, result interface{}) error {
+	alert.cfg.Elasticsearch.Endpoint.Path = path.Join(urlAttributes)
+	requestUrl := alert.cfg.Elasticsearch.Endpoint.String()
+	payload := strings.NewReader(payloadFormated)
+	alert.log.V(3).Info(fmt.Sprintf("creating HTTP %s request on %s", httpMethod, requestUrl))
+	req, err := http.NewRequest(httpMethod, requestUrl, payload)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create a http request for %s", requestUrl)
+	}
+
+	req.SetBasicAuth(alert.cfg.Elasticsearch.User, alert.cfg.Elasticsearch.Pass)
+	req.Header.Add("Content-Type", "application/json")
+	req.Header.Add("Cache-Control", "no-cache")
+
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return errors.Wrap(err, "failed to execute HTTP request")
+	}
+	defer res.Body.Close()
+
+	body, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return errors.Wrap(err, "failed to read response body")
+	}
+	if result != nil {
+		if err := json.Unmarshal(body, result); err != nil {
+			return errors.Wrapf(err, "failed to unmarshal %s", string(body))
+		}
+	}
+	return nil
+}
+
+//filePostedAlerts posts test contexts to elasticsearch
+func (alert *Alert) filePostedAlerts(tests map[string]TestDetails) error {
+	payload := generatePostedAlertsPayload(tests)
+	if err := alert.elasticRequest("/tm-alert/_doc/_bulk", "POST", payload, nil); err != nil {
+		return errors.Wrap(err, "failed to store alerted tests in elasticsearch")
+	}
+	alert.log.V(3).Info(fmt.Sprintf("filed %d tests as alerted in elasticsearch", len(tests)))
+	return nil
+}
+
+func (alert *Alert) extractRecoveredTests(testContextToTestMap map[string]TestDetails, filedAlerts alertDocs) map[string]TestDetails {
+	recoveredTests := make(map[string]TestDetails)
+	for _, filedAlert := range filedAlerts.Hits.AlertItems {
+		test, ok := testContextToTestMap[filedAlert.Source.TestName]
+		if ok && test.Successful {
```

Comment:
this recovery is only true for `last x failed` but not if the alert was fired because of a success rate. So I think we should consider both cases and implement different recovery strategies

Reply:
actually this variable considers both cases. Maybe the variable name is misleading. If you have a better naming proposal I would gladly apply it.

Reply:
ah ok then this is great 👍